0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1181 @@
+| `additional_info`      | Object               | _(Optional)_ Arbitrary fields attached to this transfer. (For example, the IDs of related transfers in other systems.) |
+| `memo`                 | Object               | _(Optional)_ Arbitrary data provided by the `debit_account` for the `credit_account`. The connector stores the ILP Header in this field. |
+| `note_to_self`         | Object               | _(Optional)_ Arbitrary data provided by the `debit_account` for itself. This field MUST be hidden when not authenticated as the `debit_account` or an admin. |
+| `id`                   | [URL][]              | _(Server-provided)_ Primary resource identifier for this transfer. MUST be an HTTP(S) URL where you can [get the transfer resource][Get Transfer]. |
+| `fulfillment`          | [URL][]              | (Server-provided) Path to the fulfillment for this transfer. MUST be an HTTP(S) URL where the client can [submit the fulfillment][Submit Fulfillment] or [get the fulfillment][Get Transfer Fulfillment]. MUST be provided if and only if this transfer has an `execution_condition`. |
+| `transfer_rejection`   | [URL][]              | (Server-provided) Path where a client can [reject this transfer][Reject Transfer]. MUST be an HTTP(S) URL. |
+| `rejection_reason`     | String               | _(Server-provided)_ The reason the transfer was rejected. MUST appear if and only if `state` is `rejected`. |
+| `state`                | String               | _(Server-provided)_ The current state of the transfer. Valid states are `prepared`, `executed`, and `rejected`. |
+| `timeline`             | Object               | _(Server-provided)_ Timeline of the transfer's state transitions. |
+| `timeline.executed_at` | [Date-Time][]        | _(Server-provided)_ Time when the transfer was originally executed. MUST appear if and only if `state` is `executed`. This time MUST be equal to or later than the `prepared_at` time. |
+| `timeline.prepared_at` | [Date-Time][]        | _(Server-provided)_ Time when the transfer was originally prepared. MUST appear, even if the transaction is unconditional. |
+| `timeline.rejected_at` | [Date-Time][]        | _(Server-provided)_ Time when the transfer was originally rejected. MUST appear if and only if `state` is `rejected`. This time MUST be equal to or later than the `prepared_at` time. |
+
+[UUID]: http://en.wikipedia.org/wiki/Universally_unique_identifier
+
+
+#### Client IDs
+
+The transfer's `client_id` field is one of the main ways of identifying the transfer; it is unique across the ledger, so implementations MAY use it as a primary unique key in a database. The client applications, not the server, generate the `client_id`, so the server SHOULD require that the field match the canonical form for a [UUID][], which is 32 lowercase hexadecimal digits, separated by hyphens into groups of 8-4-4-4-12.
```

Comment:
What do you guys think about having the ledger generate this identifier value instead of the client?  

There are quite a few requirements for this identifier that should probably be the burden of the ledger since the ledger is the one mandating them -- e.g., unique across the ledger; conforming to a particular pattern that the ledger can use as a PK in the ledger's database, etc.  Plus, the point about preventing id-squatting seems like something that should be the responsibility of the ledger as opposed to the client (who might mess it up or be unable to enforce the requirement).

For my proposal to work, though, the API would need to do something like the following to facilitate the creation of a transfer, which makes it a bit more chatty, but IMHO makes the API more robust: 

1. POST the transfer payload (minus client_id) to `/transfers`
1. Receive the created transfer as a response, with a server-supplied transfer `id`.
1. When ready, execute a PUT to `/transfers/{transfer_id}` (or possibly a PATCH) to update the state of the transfer to indicate to the ledger that the transfer is ready to be acted upon (e.g., update the state from "proposed" to "prepared", or something like that).

While the second call requiring clients to overtly trigger the processing of a transfer is somewhat of a negative, this two-phase creation has the added benefit of allowing clients to operate upon the server-side state of a transfer using possibly multiple remote calls before actually having the ledger do anything with it.